<a href="https://colab.research.google.com/github/riskakurniaa/Machine-Learning/blob/Jobsheet6/Tugas_Ensemble_Learning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Nama  : **Riska Kurnia Triwulandari**  
Kelas : **3A**  
NIM   : **2241720039**

### **Tugas 1**

Terdapat dataset mushroom. Berdasarkan dataset yang tersebut, bandingkan peforma antara algoritma Decision Tree dan RandomForest. Gunakan tunning hyperparameter untuk mendapatkan parameter dan akurasi yang terbaik.

In [19]:
# Load data
df = pd.read_csv('mushrooms.csv')
df.head()

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0     p         x           s         n       t    p               f   
1     e         x           s         y       t    a               f   
2     e         b           s         w       t    l               f   
3     p         x           y         w       t    p               f   
4     e         x           s         g       f    n               f   

  gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0            c         n          k  ...                        s   
1            c         b          k  ...                        s   
2            c         b          n  ...                        s   
3            c         n          n  ...                        s   
4            w         b          k  ...                        s   

  stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                      w                      w         p          w   
1                      w                      w         p          w   
2                      w                      w         p          w   
3                      w                      w         p          w   
4                      w                      w         p          w   

  ring-number ring-type spore-print-color population habitat  
0           o         p                 k          s       u  
1           o         p                 n          n       g  
2           o         p                 n          n       m  
3           o         p                 k          s       u  
4           o         e                 n          a       g  

[5 rows x 23 columns]

In [3]:
# Cek kolom null
df.isnull().sum()

class                       0
cap-shape                   0
cap-surface                 0
cap-color                   0
bruises                     0
odor                        0
gill-attachment             0
gill-spacing                0
gill-size                   0
gill-color                  0
stalk-shape                 0
stalk-root                  0
stalk-surface-above-ring    0
stalk-surface-below-ring    0
stalk-color-above-ring      0
stalk-color-below-ring      0
veil-type                   0
veil-color                  0
ring-number                 0
ring-type                   0
spore-print-color           0
population                  0
habitat                     0
dtype: int64

In [4]:
label_encoder = LabelEncoder()

# Loop melalui semua kolom dalam DataFrame
for column in df.columns:
    # Periksa apakah tipe data kolom adalah objek (data kategorikal)
    if df[column].dtype == 'object':
        df[column] = label_encoder.fit_transform(df[column])

# cek data yang sudah diencode
df.head()

class  cap-shape  cap-surface  cap-color  bruises  odor  gill-attachment  \
0      1          5            2          4        1     6                1   
1      0          5            2          9        1     0                1   
2      0          0            2          8        1     3                1   
3      1          5            3          8        1     6                1   
4      0          5            2          3        0     5                1   

   gill-spacing  gill-size  gill-color  ...  stalk-surface-below-ring  \
0             0          1           4  ...                         2   
1             0          0           4  ...                         2   
2             0          0           5  ...                         2   
3             0          1           5  ...                         2   
4             1          0           4  ...                         2   

   stalk-color-above-ring  stalk-color-below-ring  veil-type  veil-color  \
0                       7                       7          0           2   
1                       7                       7          0           2   
2                       7                       7          0           2   
3                       7                       7          0           2   
4                       7                       7          0           2   

   ring-number  ring-type  spore-print-color  population  habitat  
0            1          4                  2           3        5  
1            1          4                  3           2        1  
2            1          4                  3           2        3  
3            1          4                  2           3        5  
4            1          0                  3           0        1  

[5 rows x 23 columns]

In [5]:
# seleksi fitur
X = df.iloc[:,2:]
y = df['class']

# Cek jumlah fitur dan instance
X.shape

# encode label
ec = LabelEncoder()
y = ec.fit_transform(y)

# Cek jumlah fitur dan instance
print(X.shape)

# Cek label
print(y)

(8124, 21)
[1 0 0 ... 0 1 0]


In [6]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)

### Decision Tree

In [7]:
from sklearn.model_selection import GridSearchCV

dt = DecisionTreeClassifier()

params = {'max_depth': [None, 10, 20, 30],
          'min_samples_split': [2, 5, 10],
          'min_samples_leaf': [1, 2, 4]}

GS = GridSearchCV(estimator=dt,param_grid=params,cv=5,n_jobs=-1, verbose=True, scoring='accuracy')

# training/fit
GS.fit(X_train, y_train)

# mengambil model Decision Tree terbaik yang telah dihasilkan oleh proses Grid Search.
best_dtree = GS.best_estimator_

# Memprediksi label set test
y_pred_dt = best_dtree.predict(X_test)

#  menghitung set accuracy
acc_dt = accuracy_score(y_test, y_pred_dt)
print("Accuracy    : {} %".format(round((acc_dt*100),2)))
print("Test set accuracy: {:.2f}".format(acc_dt))
print(f"Test set accuracy: {acc_dt}")

Fitting 5 folds for each of 36 candidates, totalling 180 fits
Accuracy    : 100.0 %
Test set accuracy: 1.00
Test set accuracy: 1.0


### RandomForest

In [8]:
rf = RandomForestClassifier(n_estimators=10, random_state=1, n_jobs=-1)

# Sesuaikan dt ke set training
rf.fit(X_train, y_train)

# Memprediksi label set test
y_pred_rf = rf.predict(X_test)

#  menghitung set accuracy
acc_rf = accuracy_score(y_test, y_pred_rf)
print("Accuracy    : {} %".format(round((acc_rf*100),2)))
print("Test set accuracy: {:.2f}".format(acc_rf))
print(f"Test set accuracy: {acc_rf}")

Accuracy    : 100.0 %
Test set accuracy: 1.00
Test set accuracy: 1.0


###**Kesimpulan:**

Kedua model, Decision Tree dan Random Forest, menunjukkan performa yang sangat baik dengan akurasi sempurna pada dataset mushroom dengan hasil yang sama yaitu 100%.

### **Tugas 2**

Terdapat dataset mushroom. Berdasarkan dataset tersebut, bandingkan peforma antara algoritma Decision Tree dan AdaBoost. Gunakan tunning hyperparameter untuk mendapatkan parameter dan akurasi yang terbaik.

In [9]:
import numpy as np
import pandas as pd
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report
from sklearn.preprocessing import LabelEncoder

In [20]:
df2 = pd.read_csv('mushrooms.csv')
df2.head()

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0     p         x           s         n       t    p               f   
1     e         x           s         y       t    a               f   
2     e         b           s         w       t    l               f   
3     p         x           y         w       t    p               f   
4     e         x           s         g       f    n               f   

  gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0            c         n          k  ...                        s   
1            c         b          k  ...                        s   
2            c         b          n  ...                        s   
3            c         n          n  ...                        s   
4            w         b          k  ...                        s   

  stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                      w                      w         p          w   
1                      w                      w         p          w   
2                      w                      w         p          w   
3                      w                      w         p          w   
4                      w                      w         p          w   

  ring-number ring-type spore-print-color population habitat  
0           o         p                 k          s       u  
1           o         p                 n          n       g  
2           o         p                 n          n       m  
3           o         p                 k          s       u  
4           o         e                 n          a       g  

[5 rows x 23 columns]

In [11]:
# Cek kolom null
df2.isnull().sum()

class                       0
cap-shape                   0
cap-surface                 0
cap-color                   0
bruises                     0
odor                        0
gill-attachment             0
gill-spacing                0
gill-size                   0
gill-color                  0
stalk-shape                 0
stalk-root                  0
stalk-surface-above-ring    0
stalk-surface-below-ring    0
stalk-color-above-ring      0
stalk-color-below-ring      0
veil-type                   0
veil-color                  0
ring-number                 0
ring-type                   0
spore-print-color           0
population                  0
habitat                     0
dtype: int64

In [12]:
label_encoder = LabelEncoder()

# Loop melalui semua kolom dalam DataFrame
for column in df.columns:
    # Periksa apakah tipe data kolom adalah objek (data kategorikal)
    if df2[column].dtype == 'object':
        df2[column] = label_encoder.fit_transform(df2[column])

# cek data setelah diencode
df2.head()

class  cap-shape  cap-surface  cap-color  bruises  odor  gill-attachment  \
0      1          5            2          4        1     6                1   
1      0          5            2          9        1     0                1   
2      0          0            2          8        1     3                1   
3      1          5            3          8        1     6                1   
4      0          5            2          3        0     5                1   

   gill-spacing  gill-size  gill-color  ...  stalk-surface-below-ring  \
0             0          1           4  ...                         2   
1             0          0           4  ...                         2   
2             0          0           5  ...                         2   
3             0          1           5  ...                         2   
4             1          0           4  ...                         2   

   stalk-color-above-ring  stalk-color-below-ring  veil-type  veil-color  \
0                       7                       7          0           2   
1                       7                       7          0           2   
2                       7                       7          0           2   
3                       7                       7          0           2   
4                       7                       7          0           2   

   ring-number  ring-type  spore-print-color  population  habitat  
0            1          4                  2           3        5  
1            1          4                  3           2        1  
2            1          4                  3           2        3  
3            1          4                  2           3        5  
4            1          0                  3           0        1  

[5 rows x 23 columns]

In [13]:
# seleksi fitur
X = df2.iloc[:,2:]
y = df2['class']

# Cek jumlah fitur dan instance
X.shape

# encode label
ec = LabelEncoder()
y = ec.fit_transform(y)

# Cek jumlah fitur dan instance
print(X.shape)

# Cek label
print(y)

(8124, 21)
[1 0 0 ... 0 1 0]


In [14]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)

### Decision Tree

In [15]:
from sklearn.model_selection import GridSearchCV

dt = DecisionTreeClassifier()

params = {'max_depth': [None, 10, 20, 30],
          'min_samples_split': [2, 5, 10],
          'min_samples_leaf': [1, 2, 4]}

GS = GridSearchCV(estimator=dt,param_grid=params,cv=5,n_jobs=-1, verbose=True, scoring='accuracy')

# training/fit
GS.fit(X_train, y_train)

# mengambil model Decision Tree terbaik yang telah dihasilkan oleh proses Grid Search.
best_dtree = GS.best_estimator_

# Memprediksi label set test
y_pred_dt = best_dtree.predict(X_test)

#  menghitung set accuracy
acc_dt = accuracy_score(y_test, y_pred_dt)
print("Accuracy    : {} %".format(round((acc_dt*100),2)))
print("Test set accuracy: {:.2f}".format(acc_dt))
print(f"Test set accuracy: {acc_dt}")

Fitting 5 folds for each of 36 candidates, totalling 180 fits
Accuracy    : 100.0 %
Test set accuracy: 1.00
Test set accuracy: 1.0


### AdaBoost

In [16]:
ada = AdaBoostClassifier(n_estimators=2)

# Sesuaikan dt ke set training
ada.fit(X_train, y_train)

# Memprediksi label set test
y_pred_ada = ada.predict(X_test)

#  menghitung set accuracy
acc_ada = accuracy_score(y_test, y_pred_ada)
print("Test set accuracy: {:.2f}".format(acc_ada))
print(f"Test set accuracy: {acc_ada:.2f}")

Test set accuracy: 0.84
Test set accuracy: 0.84


/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_weight_boosting.py:527: FutureWarning: The SAMME.R algorithm (the default) is deprecated and will be removed in 1.6. Use the SAMME algorithm to circumvent this warning.
  warnings.warn(


### **Kesimpulan:**

Decision Tree menunjukkan performa optimal dengan akurasi 100% pada dataset ini, sementara AdaBoost memberikan akurasi 84%. Meskipun AdaBoost adalah teknik yang kuat untuk meningkatkan performa model, dalam kasus ini, ia tidak mencapai tingkat akurasi yang sama seperti Decision Tree..

### **Tugas 3**

Dengan menggunakan dataset diabetes, buatlah ensemble voting dengan algoritma

1. Logistic Regression
2. SVM kernel polynomial
3. Decission Tree

Anda boleh melakukan eksplorasi dengan melakukan tunning hyperparameter

In [17]:
import numpy as np
import pandas as pd
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.ensemble import VotingClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report

In [22]:
# Load Data
dbt = pd.read_csv('diabetes.csv')
dbt.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [23]:
# Cek nama kolom
dbt.columns

Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')

In [24]:
# Cek kolom null
dbt.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

In [26]:
from sklearn.impute import SimpleImputer

# melakukakan imputasi kolom yang memiliki data sama dengan 0
feature_columns = ['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI', 'DiabetesPedigreeFunction', 'Age']
fill_values = SimpleImputer(missing_values=0, strategy="mean", copy=False)
dbt[feature_columns] = fill_values.fit_transform(dbt[feature_columns])

In [27]:
# Split data dengan ratio 70:30
X = dbt[feature_columns]
y = dbt.Outcome

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

### Split data training dan data testing

In [28]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()

# Standarisasi pada fitur di X_train dan X_test
X_train_std = sc.fit_transform(X_train)
X_test_std = sc.transform(X_test)

### Training dengan logistic regression

In [36]:
# Buat objek Logistic Regression
lr = LogisticRegression(solver='liblinear')

# Hyperparameter grid
param_grid = {
    'C': [0.001, 0.01, 0.1, 1, 10],
    'penalty': ['l1', 'l2'],
    'max_iter': [100, 500, 1000]
}

# Penyetelan hyperparameter menggunakan Grid Search
grid_search_lr = GridSearchCV(estimator=lr, param_grid=param_grid, cv=5, scoring='accuracy')
grid_search_lr.fit(X_train_std, y_train)

# Evaluasi Logistic Regression dengan hyperparameter terbaik
best_lr = grid_search_lr.best_estimator_

# Prediksi data testing
y_pred_lr = best_lr.predict(X_test_std)

# Evaluasi akurasi data testing
acc_lr = accuracy_score(y_test, y_pred_lr)

# Print hasil evaluasi
print("Accuracy    : {} %".format(round((acc_lr*100),2)))
print("Test set accuracy: {:.2f}".format(acc_lr))
print(f"Test set accuracy: {acc_lr}")

Accuracy    : 73.59 %
Test set accuracy: 0.74
Test set accuracy: 0.7359307359307359


### Training dengan svm polynomial

In [33]:
# model svm polynomial
svm_poly = SVC(kernel='poly')

# hyperparameter
param_grid = {
    'C': [0.1, 1, 10],
    'degree': [2, 3, 4],
    'coef0': [0.0, 1.0, 2.0],
}

# penyetelan hyperparameter menggunakan grid search
grid_search_svm_poly = GridSearchCV(estimator=svm_poly, param_grid=param_grid, cv=5, scoring='accuracy')
grid_search_svm_poly.fit(X_train_std, y_train)

# evaluasi svm polynomial dengan hyperparameter terbaik
best_svm_poly = grid_search_svm_poly.best_estimator_

# predeksi data testing
y_pred_svm_poly = best_svm_poly.predict(X_test_std)

# Evaluasi akurasi data testing
acc_svm_poly = accuracy_score(y_test, y_pred_svm_poly)

# Print hasil evaluasi
print("Accuracy    : {} %".format(round((acc_svm_poly*100),2)))
print("Test set accuracy: {:.2f}".format(acc_svm_poly))
print(f"Test set accuracy: {acc_svm_poly}")

Accuracy    : 74.46 %
Test set accuracy: 0.74
Test set accuracy: 0.7445887445887446


### Training dengan decission tree

In [34]:
dt = DecisionTreeClassifier()

params = {'max_depth': [None, 10, 20, 30],
          'min_samples_split': [2, 5, 10],
          'min_samples_leaf': [1, 2, 4]}

GS = GridSearchCV(estimator=dt,param_grid=params,cv=5,n_jobs=-1, verbose=True, scoring='accuracy')

# training/fit
GS.fit(X_train_std, y_train)

# mengambil model Decision Tree terbaik yang telah dihasilkan oleh proses Grid Search.
best_dtree = GS.best_estimator_

# Memprediksi label set test
y_pred_dt = best_dtree.predict(X_test_std)

#  menghitung set accuracy
acc_dt = accuracy_score(y_test, y_pred_dt)
print("Accuracy    : {} %".format(round((acc_dt*100),2)))
print("Test set accuracy: {:.2f}".format(acc_dt))
print(f"Test set accuracy: {acc_dt}")

Fitting 5 folds for each of 36 candidates, totalling 180 fits
Accuracy    : 71.0 %
Test set accuracy: 0.71
Test set accuracy: 0.70995670995671


### Training dengan voting

In [37]:
# Definisikan algoritma yang akan digunakan untuk voting
clf1 = best_lr
clf2 = best_svm_poly
clf3 = best_dtree

# model hard voting
voting = VotingClassifier(estimators=[('logistic_reg', clf1), ('SVM-POLY', clf2), ('decission_tree', clf3)], voting='hard')

# Fit model
voting.fit(X_train_std, y_train)

# Prediksi
y_pred_vt1 = voting.predict(X_test_std)

# Evaluasi akurasi testing data
acc_vt1 = accuracy_score(y_test, y_pred_vt1)

# Print hasil evaluasi
print('Voting dengan Mayoritas')
print("Test set accuracy: {:.2f}".format(acc_vt1))
print(f"Test set accuracy: {acc_vt1}")

Voting dengan Mayoritas
Test set accuracy: 0.74
Test set accuracy: 0.7402597402597403


### **Kesimpulan**

Meskipun masing-masing model memiliki performa yang baik, penggunaan Voting Classifier yang menggabungkan hasil dari Logistic Regression, SVM dengan kernel polynomial, dan Decision Tree menghasilkan akurasi yang cukup baik yaitu 74.03%.